### PR TITLE
[worker, example] feat: add DeepSpeed engine and SFT example

### PR DESCRIPTION
**Pull request by NenoL2001**
### What does this PR do?

This PR introduces a complete implementation of a DeepSpeedEngine for VERL, providing an alternative to the existing FSDP engine. It includes the core engine logic, associated configuration data classes, and a comprehensive SFT (Supervised Fine-Tuning) example script to demonstrate its usage and validate its correctness.

The new engine supports standard training workflows, leveraging DeepSpeed's ZeRO optimizations. The example script is self-contained, handling data loading, training, metric logging, and visualization of results.

### Checklist Before Starting

- [*] Search for similar PRs. Paste at least one query link here: ...
- [ *] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test
The primary validation for this change is the end-to-end example script [deepspeed_sft_test.py]. This script runs a full training loop on a real dataset and generates a loss curve plot and a JSON summary of the training results, confirming the engine's correctness.

### API and Usage Example
python examples/engine_sft_demo/deepspeed_sft_test.py

### Design & Code Changes

New Engine: Added engine_impl and DeepSpeedEngineWithLMHead for deep speed

New config: Updated engine.py to include DeepSpeedEngineConfig and DeepSpeedOptimizerConfig

New Utilities: Added deepspeed_utils.py and deepspeed_checkout_manager.py for DeepSpeed-specific helper functions and checkpointing

Example Script: Added deepspeed_sft_test.py  and fsdp_sft_test.py as a comprehensive, runnable example and validation tool.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [* ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [*] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [*] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [*] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why:   
- [] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

**Comment by vermouth1992** [comment]
Thanks for the great contribution! Here are some general comments:
- The engine implementation is not fully ready yet. There are still some code redundancy between engines implementations (e.g., micro batch preparation, multi-model handling). And we are still finalizing them.
- It seems that DeepSpeed implementation shares most code of forward_step compared with FSDP backend, is that possible to inherit FSDP backend so that our FSDP backend modification will automatically apply in DeepSpeed backend? Later on, when more features are integrated and forward_step is significantly different, we can write DeepSpeed's own forward_step function.
- Could you add a test to here: https://github.com/volcengine/verl/blob/main/tests/special_e2e/sft/run_sft_engine_gsm8k.sh that runs GSM8k SFT with various backend, and invoke the tests here: https://github.com/volcengine/verl/blob/main/tests/special_e2e/sft/test_sft_engine_all.sh. It will run the sft trainer and compare the loss/grad_norm of the first step to ensure that the result matches with various backend and various parallelism configurations

**Comment by NenoL2001** [comment]
> Thanks for the great contribution! Here are some general comments:
> 
> * The engine implementation is not fully ready yet. There are still some code redundancy between engines implementations (e.g., micro batch preparation, multi-model handling). And we are still finalizing them.
> * It seems that DeepSpeed implementation shares most code of forward_step compared with FSDP backend, is that possible to inherit FSDP backend so that our FSDP backend modification will automatically apply in DeepSpeed backend? Later on, when more features are integrated and forward_step is significantly different, we can write DeepSpeed's own forward_step function.
> * Could you add a test to here: https://github.com/volcengine/verl/blob/main/tests/special_e2e/sft/run_sft_engine_gsm8k.sh that runs GSM8k SFT with various backend, and invoke the tests here: https://github.com/volcengine/verl/blob/main/tests/special_e2e/sft/test_sft_engine_all.sh. It will run the sft trainer and compare the loss/grad_norm of the first step to ensure that the result matches with various backend and various parallelism configurations

Thanks for the great suggestions! I have added the requested tests to ensure the results match. I've also refactored the DeepSpeed engine implementation to align with the latest base engine, which should address the code redundancy concern. Please let me know what you think.

**Comment by NenoL2001** [comment]
> Thanks for the great contribution! Here are some general comments:
> 
> * The engine implementation is not fully ready yet. There are still some code redundancy between engines implementations (e.g., micro batch preparation, multi-model handling). And we are still finalizing them.
> * It seems that DeepSpeed implementation shares most code of forward_step compared with FSDP backend, is that possible to inherit FSDP backend so that our FSDP backend modification will automatically apply in DeepSpeed backend? Later on, when more features are integrated and forward_step is significantly different, we can write DeepSpeed's own forward_step function.
> * Could you add a test to here: https://github.com/volcengine/verl/blob/main/tests/special_e2e/sft/run_sft_engine_gsm8k.sh that runs GSM8k SFT with various backend, and invoke the tests here: https://github.com/volcengine/verl/blob/main/tests/special_e2e/sft/test_sft_engine_all.sh. It will run the sft trainer and compare the loss/grad_norm of the first step to ensure that the result matches with various backend and various parallelism configurations

As a follow-up question, I've implemented the deepspeedWorker based on the fsdpWorker, though I haven't validated it on multi-GPU yet. I saw the comments about a future refactor. Is there an expected timeline for that? I'm concerned about the maintenance cost of adding a new worker if it will be refactored soon.

**Comment by NenoL2001** [comment]
Hi @vermouth1992,

Hope you are having a good week.

Following up on your earlier review. I have completed a full rewrite of the DeepSpeed worker to address your feedback. Specifically:

RL support: To support RL in DeepSpeed, I have written DeepspeedWorker and benchmarks with FSDP on Qwen3 0.5B and the GSM8K dataset under PPO.

Code Redundancy: I found DeepSpeed could not inherit the FSDP class because DeepSpeed has its own optimizer.

Testing: I have added the SFT tests as you suggested to validate the implementation against various backends and parallelism configurations.

I've conducted a detailed performance validation against the existing FSDP implementation. The results confirm that the performance issues are resolved, and the new DeepSpeed worker is ready for review.

Benchmark Results

The new DeepSpeed worker's performance is now on par with the FSDP implementation. The reward and loss curves (attached below) track very closely, indicating correct and stable training behavior across various multi-GPU configurations.

Benchmark Configurations

The benchmarks were run on the Qwen3 0.5B model with the GSM8K dataset. The attached graphs correspond to these four multi-GPU test configurations from the metrics_summary.md:

Common Settings (All Runs):

PPO Mini-batch Size: 128

PPO Micro-batch Size (per GPU): 8

Gradient Accumulation Steps: 8

Precision: bf16

ZeRO Stage: 0

Run 1: 2-GPU Data Parallel (dp_2gpu_full)

GPUs: 2

Data Parallel (DP) Size: 2

Tensor Parallel (TP) Size: 1

Final Throughput: ~1786 tokens/sec

Run 2: 2-GPU Tensor Parallel (tp_2gpu_full)

GPUs: 2

Data Parallel (DP) Size: 1

Tensor Parallel (TP) Size: 2

Final Throughput: ~1403 tokens/sec

Run 3: 4-GPU Hybrid Parallel (dp_tp_ep1 - 1 Epoch)

GPUs: 4

Data Parallel (DP) Size: 2

Tensor Parallel (TP) Size: 2

Final Throughput: ~963 tokens/sec

Run 4: 4-GPU Hybrid Parallel (dp_tp_ep2 - 2 Epochs)

GPUs: 4

Data Parallel (DP) Size: 2

Tensor Parallel (TP) Size: 2

Final Throughput: ~1396 tokens/sec

Given the active development in the main branch, I'm concerned that a delay might lead to significant merge conflicts. I would be very grateful if you could take a look when you have a moment.

Thanks for your time and guidance.

Best regards,

Qinwen Ling

metrics_progress_dp_2gpu_full metrics_progress_tp_2gpu_full metrics_progress_dp_tp_ep1 metrics_progress_dp_tp_ep2

**Comment by ISEEKYAN** [comment]
maybe we should start a new PR with a clean `verl/verl/workers/engine/deepspeed/transformer_impl.py`?
